sdk/servicebus/azservicebus/receiver.go

```diff
@@ -14,11 +24,306 @@ const (
+
+// ReceiveWithMaxWaitTime configures how long to wait for the first
+// message in a set of messages to arrive.
+// Default: 60 seconds
+func ReceiveWithMaxWaitTime(max time.Duration) ReceiveOption {
+	return func(options *ReceiveOptions) error {
+		options.maxWaitTime = max
+		return nil
+	}
+}
+
+// ReceiveWithMaxTimeAfterFirstMessage confiures how long, after the first
+// message arrives, to wait before returning.
+// Default: 1 second
+func ReceiveWithMaxTimeAfterFirstMessage(max time.Duration) ReceiveOption {
+	return func(options *ReceiveOptions) error {
+		options.maxWaitTimeAfterFirstMessage = max
+		return nil
+	}
+}
+
+// ReceiveMessages receives a fixed number of messages, up to numMessages.
+// There are two timeouts involved in receiving messages:
+// 1. An explicit timeout set with `ReceiveWithMaxWaitTime` (default: 60 seconds)
+// 2. An implicit timeout (default: 1 second) that starts after the first
+//    message has been received. This time can be adjusted with `ReceiveWithMaxTimeAfterFirstMessage`
+func (r *Receiver) ReceiveMessages(ctx context.Context, numMessages int, options ...ReceiveOption) ([]*ReceivedMessage, error) {
+	if r.linkState.Closed() {
+		return nil, r.linkState.Err()
+	}
+
+	ropts := &ReceiveOptions{
+		maxWaitTime:                  time.Minute,
+		maxWaitTimeAfterFirstMessage: time.Second,
+	}
+
+	for _, opt := range options {
+		if err := opt(ropts); err != nil {
+			return nil, err
+		}
+	}
+
+	receiver, err := r.updateReceiver(ctx)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if err := receiver.IssueCredit(uint32(numMessages)); err != nil {
+		return nil, err
+	}
+
+	var messages []*ReceivedMessage
+	var allMessagesReceivedError = errors.New("all messages received")
+
+	receiveCh, afterFirstMessageFn := startReceiverTimer(ropts.maxWaitTime, ropts.maxWaitTimeAfterFirstMessage)
+
+	listenCtx, cancelListenCtx := context.WithCancel(ctx)
+	defer cancelListenCtx()
+
+	listenHandle := r.receiver.Listen(listenCtx, internal.HandlerFunc(func(c context.Context, legacyMessage *internal.Message) error {
+		// NOTE: the invocation of this function from the AMQP library is single-threaded so
+		// no concurrency sync is required.
+		messages = append(messages, convertToReceivedMessage(legacyMessage))
+
+		if len(messages) == numMessages {
+			return allMessagesReceivedError
+		}
+
+		if len(messages) == 1 {
+			afterFirstMessageFn()
+		}
+
+		return nil
+	}))
+
+	select {
+	case <-listenHandle.Done():
+		if listenHandle.Err() != allMessagesReceivedError {
+			err = listenHandle.Err()
+		}
+	case <-receiveCh:
+		break
+	case <-r.linkState.Done():
+		err = r.linkState.Err()
+	case <-ctx.Done():
+		err = ctx.Err()
+	}
+
+	// make sure we leave the link in a consistent state.
+	if err := r.receiver.DrainCredit(ctx); err != nil {
+		return nil, err
+	}
+
+	return messages, err
+}
+
+func (r *Receiver) CompleteMessage(ctx context.Context, message *ReceivedMessage) error {
```

Comment:
Is this the track 2 pattern, i.e. place the message handling APIs on the receiver and not the message like in track 1?

Reply:
Yes. 

The big argument at the time is that we try not to use any 'Active Record' pattern with our models/data contracts.